### DISCUSSION TITLE
Cisco ASA logs not getting parsed consistently

### DISCUSSION THREAD

**Discussion by abesinger** [Security-Onion-Solutions/securityonion#5553] [Unsupported Versions]
I just turned on collection of Cisco ASA logs using the filebeatCisco ASA module. I'm seeing both inconsistent and incorrect parsing of the logs.  Details below. Has anyone seen/dealt with any of these?  I'd like to see if someone else has solved it before delving into the source code and wading through ingest pipelines to figure it out.

I suspect some of the issues are with the filebeat module itself, but I think a few (like some of the hostname issues) are specific to Security Onion:

The problems are...

Inconsistent fields:
* For some messages, the `log.original` field gets populated with the raw message, but `event.original` and `event.action` are empty.  For others, `event.original` and `event.action` are populated, but not `log.original`. I do not see anything obvious in the message contents that would indicate why that's happening. The one thing that might be a clue is that the ones with `event.original` populated also have `error.message` showing `dynamic method [java.lang.String, entrySet/0] not found.` I suspect that means a Java version mismatch, but I'm not a Java guru.

Incorrect hostnames:
*  The document doesn't have any field that identifies which host originated the message. `agent.hostname` is set to `so-filebeat` in every message.  This is just b0rked. What's the use of a log message when you don't know what host it came from? 
* `log.source.address` is set to the IP address of the container, which I think explains the previous bullet.  Still b0rked.
* `agent.name` has the (actual) hostname of the host that filebeat is running on, which is totally wrong (should be `filebeat`, no?),  but also makes the previous two bullets look even more b0rked, because filebeat obviously knows the name of the host it's running on.
* `metadata.ip_address` has the IP address of the filebeat host. I guess that's correct, but still makes the previous bullets look stupid.

Timezones in the twilight zone:
* `event.timezone` is set to `+00:00` even though the original event has a timezone in the timestamp (`EDT`). 
* `@timestamp` is not the timestamp from the log message,and `event.ingested` is neither the log timestamp nor `@timestamp`.
*  Here's an example from one message: 
 * Original timestamp: "Sep 16 16:14:17 EDT." 
 * `@timestamp`: "Sep 16, 2021 @ 15:14:17.950," 
 * event.ingested`: "2021-09-16T20:14:18.476447362Z"

Incorrect syslog facility (not properly parsing syslog priority)
* `syslog.facility` is populated with the syslog priority (`(facility << 3) & level`), not the facility.

Here are a couple of example document; I've excised fields that aren't important to see, and names have been changed to protect the innocent.  Note that @timestamp is being shown in CDT, as that's the time zone I'm viewing this stuff in, and which is why it shows as an hour earlier than the timestamp in the original message (in EDT). But that doesn't explain why @timestamp has microseconds when the original message doesn't.

FIELD | VALUE
------------ | -------------
`@timestamp` | Sep 16, 2021 @ 21:29:52.698
`@version` | 1
`agent.hostname` | so-filebeat
`agent.id` | f8a0e795-7d2a-45c4-b7af-ae302867be5f
`agent.name` | example15
`agent.type` | filebeat
`agent.version` | 7.13.4
`cisco.asa.destination_interface` | outside
`cisco.asa.message_id` | 305011
`cisco.asa.source_interface` | outside
`cisco.asa.suffix` | session
`destination.address` | 192.168.1.1
`destination.as.number` | 99
`destination.ip` | 192.168.1.1
`destination.port` | 64,043
`ecs.version` | 1.9.0
`event.action` | firewall-rule
`event.category` | network
`event.code` | 305,011
`event.dataset` | cisco.asa
`event.ingested` | 2021-09-17T02:29:53.046505741Z
`event.kind` | event
`event.module` | cisco
`event.original` | :Sep 16 22:29:52 EDT: %ASA-session-6-305011: Built dynamic TCP translation from outside:10.1.2.3/64043(LOCAL\nobody@example.com) to outside:192.168.1.1/64043
`event.severity` | 6
`event.timezone` | +00:00
`event.type` | info
`fileset.name` | asa
`input.type` | udp
`log.level` | informational
`log.source.address` | 172.17.0.1:56634
`metadata.beat` | filebeat
`metadata.ip_address` | 10.1.2.3
`metadata.pipeline` | filebeat-7.13.4-cisco-asa-asa-ftd-pipeline
`metadata.truncated` | false
`metadata.type` | _doc
`metadata.version` | 7.13.4
`network.iana_number` | 6
`network.transport` | tcp
`observer.egress.interface.name` | outside
`observer.ingress.interface.name` | outside
`observer.product` | asa
`observer.type` | firewall
`observer.vendor` | Cisco
`related.ip` | 10.2.3.4, 192.168.1.1
`service.type` | cisco
`source.address` | 10.2.3.4
`source.ip` | 10.2.3.4
`source.port` | 64,043
`source.user.name` | LOCAL\nobody@examle.com
`syslog.facility` | 190
`tags` | cisco-asa, forwarded, beats_input_codec_plain_applied
`type` | redis-input

Here's a document of a log message where `log.original` is populated, but `event.original` and other fields are not.  I suspect the `log.error` field gives a clue that there was a java error.

 FIELD | VALUE 
------------ | -------------
`@timestamp` | Sep 16, 2021 @ 21:59:52.431
`@version` | 1
`agent.hostname` | so-filebeat
`agent.name` | itsirlsa15
`agent.type` | filebeat
`agent.version` | 7.13.4
`cisco.asa.message_id` |
`ecs.version` | 1.9.0
`error.message` | dynamic method [java.lang.String, entrySet/0] not found
`event.dataset` | cisco.asa
`event.ingested` | 2021-09-17T02:59:53.829818853Z
`event.module` | cisco
`event.severity` | 7
`event.timezone` | +00:00
`fileset.name` | asa
`input.type` | udp
`log.level` | debug
`log.original` | :Sep 16 22:59:52 EDT: %ASA--4-419002: Duplicate TCP SYN from inside:10.1.2.3.4 to AZURE-VTI01:172.17.1.2/123 with different initial sequence number
`log.source.address` | 172.17.0.1:43096
`metadata.beat` | filebeat
`metadata.ip_address` | 10.2.3.4
`metadata.pipeline` | filebeat-7.13.4-cisco-asa-asa-ftd-pipeline
`metadata.truncated` | false
`metadata.type` | _doc
`metadata.version` | 7.13.4
`service.type` | cisco
`syslog.facility` | 188
`tags` | cisco-asa, forwarded, beats_input_codec_plain_applied
`type` | redis-input

**Comment by gebhard73** [comment]
Hi,

I've written my own parsing. May be worth a look:
https://github.com/Security-Onion-Solutions/securityonion/discussions/3547

Works fine for us.

Cheers,
Gebhard
